src/pybind/mgr/dashboard/frontend/src/app/shared/components/telemetry-notification/telemetry-notification.component.spec.ts

```diff
@@ -0,0 +1,136 @@
+  });
+
+  it('should create', () => {
+    component.ngOnInit();
+    expect(component).toBeTruthy();
+  });
+
+  it('should not show notification again if the user closed it before', () => {
+    spyOn(component, 'isNotificationHidden').and.returnValue(true);
+    fixture.detectChanges();
+    component.ngOnInit();
+    expect(component.displayNotification).toBe(false);
+  });
+
+  it('should not show notification for an user without administrator role', () => {
+    spyOn(component, 'isNotificationHidden').and.returnValue(false);
+    spyOn(authStorageService, 'getUsername').and.returnValue('username');
+    spyOn(userService, 'get').and.returnValue(of(user));
+    fixture.detectChanges();
+    component.ngOnInit();
+    expect(component.displayNotification).toBe(false);
+  });
+
+  it('should not show notification if the module is enabled already', () => {
+    spyOn(component, 'isNotificationHidden').and.returnValue(false);
+    spyOn(authStorageService, 'getUsername').and.returnValue('admin');
+    spyOn(userService, 'get').and.returnValue(of(admin));
+    spyOn(mgrModuleService, 'getConfig').and.returnValue(of(telemetryEnabledConfig));
+    fixture.detectChanges();
+    component.ngOnInit();
+    expect(component.displayNotification).toBe(false);
+  });
+
+  it('should show the notification if all pre-conditions set accordingly', () => {
+    spyOn(component, 'isNotificationHidden').and.returnValue(false);
+    spyOn(authStorageService, 'getUsername').and.returnValue('username');
+    spyOn(userService, 'get').and.returnValue(of(admin));
+    spyOn(mgrModuleService, 'getConfig').and.returnValue(of(telemetryDisabledConfig));
+    fixture.detectChanges();
+    component.ngOnInit();
+    expect(component.displayNotification).toBe(true);
+  });
+
+  it('should hide the notification if the user closes it', () => {
+    spyOn(component, 'isNotificationHidden').and.returnValue(false);
+    spyOn(authStorageService, 'getUsername').and.returnValue('username');
+    spyOn(userService, 'get').and.returnValue(of(admin));
+    spyOn(mgrModuleService, 'getConfig').and.returnValue(of(telemetryDisabledConfig));
+    spyOn(notificationService, 'show');
+    fixture.detectChanges();
+    component.ngOnInit();
+    component.close();
+    expect(notificationService.show).toHaveBeenCalled();
+    expect(localStorage.getItem('telemetry_notification_hidden')).toBe('true');
+  });
+
+  it('should hide the notification if the user loggs out', () => {
+    const telemetryNotificationService = TestBed.get(TelemetryNotificationService);
+    spyOn(component, 'isNotificationHidden').and.returnValue(false);
+    spyOn(authStorageService, 'getUsername').and.returnValue('username');
+    spyOn(userService, 'get').and.returnValue(of(admin));
+    spyOn(mgrModuleService, 'getConfig').and.returnValue(of(telemetryDisabledConfig));
+    spyOn(telemetryNotificationService, 'setVisibility');
+    fixture.detectChanges();
+    component.ngOnInit();
```

Comment:
Hi, first thanks for all those tests :D

I just refactored it a bit to create nearly all spies in the `beforeEach` and I removed the `ngOnInit` call which is not needed as `detectChanges` will trigger it on the first run.

Here is the diff:
```diff
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/shared/components/telemetry-notification/telemetry-notification.component.spec.ts b/src/pybind/mgr/dashboard/frontend/src/app/shared/components/telemetry-notification/telemetry-notification.component.spec.ts
index 7424cf53def..8db0de0dc0d 100644
--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/components/telemetry-notification/telemetry-notification.component.spec.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/components/telemetry-notification/telemetry-notification.component.spec.ts
@@ -19,11 +19,16 @@ import { TelemetryNotificationComponent } from './telemetry-notification.compone
 describe('TelemetryActivationNotificationComponent', () => {
   let component: TelemetryNotificationComponent;
   let fixture: ComponentFixture<TelemetryNotificationComponent>;
-
+  // Services
   let authStorageService: AuthStorageService;
   let userService: UserService;
   let mgrModuleService: MgrModuleService;
   let notificationService: NotificationService;
+  // Spies
+  let isNotificationHiddenSpy: jasmine.Spy;
+  let getUsernameSpy: jasmine.Spy;
+  let userServiceGetSpy: jasmine.Spy;
+  let getConfigSpy: jasmine.Spy;
 
   const user: UserFormModel = {
     username: 'username',
@@ -65,71 +70,56 @@ describe('TelemetryActivationNotificationComponent', () => {
     userService = TestBed.get(UserService);
     mgrModuleService = TestBed.get(MgrModuleService);
     notificationService = TestBed.get(NotificationService);
+    // Spies
+    isNotificationHiddenSpy = spyOn(component, 'isNotificationHidden').and.returnValue(false);
+    getUsernameSpy = spyOn(authStorageService, 'getUsername').and.returnValue('username');
+    userServiceGetSpy = spyOn(userService, 'get').and.returnValue(of(admin)); // Not the best name but it sounded better than `getSpy`
+    getConfigSpy = spyOn(mgrModuleService, 'getConfig').and.returnValue(
+      of(telemetryDisabledConfig)
+    );
   });
 
   it('should create', () => {
-    component.ngOnInit();
+    fixture.detectChanges();
     expect(component).toBeTruthy();
   });
 
   it('should not show notification again if the user closed it before', () => {
-    spyOn(component, 'isNotificationHidden').and.returnValue(true);
+    isNotificationHiddenSpy.and.returnValue(true);
     fixture.detectChanges();
-    component.ngOnInit();
     expect(component.displayNotification).toBe(false);
   });
 
   it('should not show notification for an user without administrator role', () => {
-    spyOn(component, 'isNotificationHidden').and.returnValue(false);
-    spyOn(authStorageService, 'getUsername').and.returnValue('username');
-    spyOn(userService, 'get').and.returnValue(of(user));
+    userServiceGetSpy.and.returnValue(of(user));
     fixture.detectChanges();
-    component.ngOnInit();
     expect(component.displayNotification).toBe(false);
   });
 
   it('should not show notification if the module is enabled already', () => {
-    spyOn(component, 'isNotificationHidden').and.returnValue(false);
-    spyOn(authStorageService, 'getUsername').and.returnValue('admin');
-    spyOn(userService, 'get').and.returnValue(of(admin));
-    spyOn(mgrModuleService, 'getConfig').and.returnValue(of(telemetryEnabledConfig));
+    getUsernameSpy.and.returnValue('admin');
+    getConfigSpy.and.returnValue(of(telemetryEnabledConfig));
     fixture.detectChanges();
-    component.ngOnInit();
     expect(component.displayNotification).toBe(false);
   });
 
   it('should show the notification if all pre-conditions set accordingly', () => {
-    spyOn(component, 'isNotificationHidden').and.returnValue(false);
-    spyOn(authStorageService, 'getUsername').and.returnValue('username');
-    spyOn(userService, 'get').and.returnValue(of(admin));
-    spyOn(mgrModuleService, 'getConfig').and.returnValue(of(telemetryDisabledConfig));
     fixture.detectChanges();
-    component.ngOnInit();
     expect(component.displayNotification).toBe(true);
   });
 
   it('should hide the notification if the user closes it', () => {
-    spyOn(component, 'isNotificationHidden').and.returnValue(false);
-    spyOn(authStorageService, 'getUsername').and.returnValue('username');
-    spyOn(userService, 'get').and.returnValue(of(admin));
-    spyOn(mgrModuleService, 'getConfig').and.returnValue(of(telemetryDisabledConfig));
     spyOn(notificationService, 'show');
     fixture.detectChanges();
-    component.ngOnInit();
     component.close();
     expect(notificationService.show).toHaveBeenCalled();
     expect(localStorage.getItem('telemetry_notification_hidden')).toBe('true');
   });
 
-  it('should hide the notification if the user loggs out', () => {
+  it('should hide the notification if the user logs out', () => {
     const telemetryNotificationService = TestBed.get(TelemetryNotificationService);
-    spyOn(component, 'isNotificationHidden').and.returnValue(false);
-    spyOn(authStorageService, 'getUsername').and.returnValue('username');
-    spyOn(userService, 'get').and.returnValue(of(admin));
-    spyOn(mgrModuleService, 'getConfig').and.returnValue(of(telemetryDisabledConfig));
     spyOn(telemetryNotificationService, 'setVisibility');
     fixture.detectChanges();
-    component.ngOnInit();
     component.ngOnDestroy();
     expect(telemetryNotificationService.setVisibility).toHaveBeenCalledWith(false);
   });
```

Reply:
Thank you for your suggestion! I'll adapt the tests accordingly.